test/e2e/common.go

```diff
@@ -35,15 +36,64 @@ type runOpts struct {
 	Redact       bool
 }
 
+var (
+	e env
+	k kn
+)
+
+const (
+	KnDefaultTestImage string = "gcr.io/knative-samples/helloworld-go"
+	MaxRetries         int    = 10
+)
+
+// Setup set up an enviroment for kn integration test returns the Teardown cleanup function
+func Setup(t *testing.T) func(t *testing.T) {
+	e = buildEnv(t)
+	e.Namespace = fmt.Sprintf("%s%d", e.Namespace, namespaceCount)
+	namespaceCount++
+	k = kn{t, e.Namespace, Logger{}}
+	CreateTestNamespace(t, e.Namespace)
+	return Teardown
+}
+
+// Teaddown clean up
+func Teardown(t *testing.T) {
+	DeleteTestNamespace(t, e.Namespace)
+}
+
 // CreateTestNamespace creates and tests a namesspace creation invoking kubectl
 func CreateTestNamespace(t *testing.T, namespace string) {
-	kubectl := kubectl{t, Logger{}}
-	out, err := kubectl.RunWithOpts([]string{"create", "namespace", namespace}, runOpts{})
-	if err != nil {
-		t.Fatalf(fmt.Sprintf("Error executing 'kubectl create namespace' command. Error: %s", err.Error()))
+	logger := Logger{}
+	kubectlCreateNamespace := func() (string, error) {
+		kubectl := kubectl{t, logger}
+		return kubectl.RunWithOpts([]string{"create", "namespace", namespace}, runOpts{AllowError: true})
 	}
 
 	expectedOutputRegexp := fmt.Sprintf("namespace?.+%s.+created", namespace)
+	var (
+		created bool
+		retries int
+		out     string
+		err     error
+	)
+
+	for !created && retries < MaxRetries {
+		out, err = kubectlCreateNamespace()
+		if err != nil && retries >= MaxRetries {
```

Comment:
Can we use an assert.NotNil() here ?

Reply:
Either one but I prefer it in the logic since I am not testing here.